### DISCUSSION TITLE
`vercel dev` fails during preflight

### DISCUSSION THREAD

**Discussion by devatina11yb** [vercel/community#2173] [Help] [answered]
### Summary

I get a `TypeError: argument string is required` during preflight while serving locally with `vercel dev`. I believe it is because of the missing Content-Type header while parsing the body of the request inside the Serverless Functions helpers.

Oddly enough, it does work without error once deployed in production.

I built a simple project to illustrate the issue.
GitHub repository: https://github.com/devatina11yb/vercel-yoga.

Development endpoint: http://localhost:3000/api.
Production endpoint: https://vercel-yoga.vercel.app/api.

To reproduce the error, you need to run `pnpm run serve:dev`.

It is a GraphQL API using GraphQL Yoga so there is a possibility that the issue comes from GraphQL Yoga but I still think it is an issue with the Serverless Functions helpers.

### Additional information

Here is the full error message:
```
...\vercel-yoga\node_modules\.pnpm\content-type@1.0.5\node_modules\content-type\index.js:108
    throw new TypeError('argument string is required')
          ^

TypeError: argument string is required
    at parse (...\vercel-yoga\node_modules\.pnpm\content-type@1.0.5\node_modules\content-type\index.js:108:11)
    at parseBody (...\vercel-yoga\node_modules\.pnpm\@vercel+node@2.12.0\node_modules\@vercel\node\dist\serverless-functions\helpers.js:15:26)
    at IncomingMessage.get [as body] (...\vercel-yoga\node_modules\.pnpm\@vercel+node@2.12.0\node_modules\@vercel\node\dist\serverless-functions\helpers.js:80:27)
    at normalizeNodeRequest (file:///.../vercel-yoga/node_modules/.pnpm/@whatwg-node+server@0.7.5/node_modules/@whatwg-node/server/esm/utils.js:76:41)
    at handleNodeRequest (file:///.../vercel-yoga/node_modules/.pnpm/@whatwg-node+server@0.7.5/node_modules/@whatwg-node/server/esm/createServerAdapter.js:82:25)
    at requestListener (file:///.../vercel-yoga/node_modules/.pnpm/@whatwg-node+server@0.7.5/node_modules/@whatwg-node/server/esm/createServerAdapter.js:96:32)
    at genericRequestHandler (file:///.../vercel-yoga/node_modules/.pnpm/@whatwg-node+server@0.7.5/node_modules/@whatwg-node/server/esm/createServerAdapter.js:155:20)
    at Object.apply (file:///.../vercel-yoga/node_modules/.pnpm/@whatwg-node+server@0.7.5/node_modules/@whatwg-node/server/esm/createServerAdapter.js:220:20)
    at Server.<anonymous> (...\vercel-yoga\node_modules\.pnpm\@vercel+node@2.12.0\node_modules\@vercel\node\dist\serverless-functions\serverless-handler.js:20:16)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
```

**Comment by amyegan** [comment]
Hi @devatina11yb! When trying to replicate the Vercel environment on local, you should use [`vercel dev`](https://vercel.com/docs/cli/dev). If I use that command, the sample project runs at `localhost:3000` without build errors. 

A `TypeError` would be thrown because of TypeScript issues. 😕 But I'm not seeing the same error when I run `pnpm run serve:dev`, so I'm not sure what the reason is exactly. Please try with `vercel dev` and it should work for you 🙏

**Comment by devatina11yb** [comment]
`pnpm run serve:dev` does call `vercel dev`, in package.json:
```json
"scripts": {
    "serve:dev": "pnpm run build && dotenv -e ./.env.development -- vercel dev"
  }
```

The error happens during preflight (OPTIONS request with headers for CORS but no Content-Type header) with [Apollo Sandbox](https://studio.apollographql.com/sandbox/explorer) for example.

It actually works with GraphiQL which is bundled with GraphQL Yoga since GraphiQL doesn’t send preflight requests.

I do believe any preflight request without Content-Type header (like any regular OPTIONS request) would fail.

**Reply by amyegan** [reply]
Everything seems to be working normally for me with `vercel dev`. Is there a specific URL I need to visit to trigger the error? Maybe I'm not fully understanding the problem 😕

**Reply by devatina11yb** [reply]
You’ll get the error if you try to fetch the GraphQL Schema or query the API at `http://localhost:3000/api` with Apollo Sandbox.

The only available query is:
```gql
query {
  knockknock
}
```

**Reply by amyegan** [reply]
When I start the project with `vercel dev`, then visit `http://localhost:3000/api` and run the query, I get a response with no errors in the browser console or my local terminal

<img src="https://user-images.githubusercontent.com/8247105/234695243-48270489-eedb-43c2-afca-fc653b34fe45.png" alt="query knock knock responded with who's there" />

**Comment by NekoCharm01** [comment]
I'm experiencing the same problem, and I'm sure that it is `@vercel/node` that results in the issue. I made a minimal reproduction: https://github.com/AlPha5130/bug-vercel-node

Steps to reproduce:
1. clone the repo
2. run `pnpm install`
3. run `pnpm vercel dev`
4. finish vercel's questions, all using default option
5. when the terminal says `> Ready! Available at http://localhost:3000`, open a browser and visit `localhost:3000/api/http`
6. you get an error in the browser `This Serverless Function has crashed.` and an error in the terminal `TypeError: argument string is required`

Analyze:
On https://github.com/vercel/vercel/blob/main/packages/node/src/serverless-functions/helpers.ts#L266-L269 , when incoming request does not sepcify `Content-Type`, the `contentType` is `undefined`, and `body` is set as `Buffer.from('')`. Then the `req.body` getter is set, and when it is read, it calls function `getBodyParser` on L30, with argument `body` and `contentType` set as above. On L33 the imported function `parseContentType` is called with argument `contentType`, whose value is `undefined`. The source code of the imported function throws an error when the argument is `undefined`, so the error occurs.

**Reply by ivandevp** [reply, marked as answer]
This should be fix now, check https://github.com/vercel/vercel/issues/9850#issuecomment-1585599684.
